### PR TITLE
Migrate to Koa v2 middleware syntax

### PR DESCRIPTION
**Pull request by nfantone**
- Upgrade koa in `devDependencies` to `^2.0.0`.
- Remove generators in favor of promise-based middleware.
- Update all unit tests.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -30,26 +27,26 @@ module.exports = function(opts) {
...
-    secret = (this.state && this.state.secret) ? this.state.secret : opts.secret;
+    secret = (ctx.state && ctx.state.secret) ? ctx.state.secret : opts.secret;
     if (!secret) {
-      this.throw(500, 'Invalid secret\n');
+      ctx.throw(500, 'Invalid secret\n');
     }
 
     try {
-      user = yield JWT.verify(token, secret, opts);
+      user = JWT.verify(token, secret, opts);
```
Comment (fl0w) [inline index.js]: this will verify synchronously, if one uses a certificate this will not do. I'll argue this needs to be async.

Comment (nfantone) [inline index.js]: I wrote it as an async call (using bluebird's promisify), but then I recall you guys asked me to "remove promisification" of JWT methods. And so I did. Maybe I misunderstood you.

Comment (stiang) [inline index.js]: Ah, sorry about that. I meant these parts in index.js (notice how `verify` gets thunkified:

```js
// Make verify function play nice with co/koa
var JWT = {decode: _JWT.decode, sign: _JWT.sign, verify: thunkify(_JWT.verify)};
```

and then later, at the bottom of the file:

```js
// Export JWT methods as a convenience
module.exports.sign   = _JWT.sign;
module.exports.verify = _JWT.verify;
module.exports.decode = _JWT.decode;
```

... which means that we export the thunkified `verify`, which is kind of messy.

This is discussed in #32.

Comment (nfantone) [inline index.js]: Done!

**Comment by fl0w** [comment]
Can I recommend you using either `jwt-then` package and `any-promise`?
`any-promise` allows consumers to define which Promise package to use themselves, and the only reason you use `bluebird` is to promisify.

If you omit `bluebird` you can instead just include `jwt-then`, which has a dependency of `jsonwebtoken` anyway.

**Comment by nfantone** [comment]
@fl0w Your library is giving me some trouble. And is the only thing preventing me from finishing this PR. I've already fixed `koa-unless`.

This,

```js
var Promise   = require('bluebird');
var JWT       = Promise.promisifyAll(require('jsonwebtoken'));
// ...
return JWT.verifyAsync(...)
```

works wonder.

```sh
$ npm test

> koa-jwt@1.2.0 test /home/nfantone/dev/js/koa/jwt
> make test

// ...
  unless tests
    ✓ should pass if the route is excluded
    ✓ should fail if the route is not excluded
    ✓ should pass if the route is not excluded and the token is present


  23 passing
```

While this,

```js
var JWT       = require('jwt-then');
// ...
return JWT.verify(...)
```

breaks unit tests.

```js
$ npm test

> koa-jwt@1.2.0 test /home/nfantone/dev/js/koa/jwt
> make test

 1) should throw if audience is not expected

  passthrough tests

  success tests

  unless tests


  7 passing (77ms)
  1 failing

  1) failure tests should throw if audience is not expected:
     Error: expected 'Invalid token - jwt audience invalid. expected: not-expected-audience\n' response body, got 'Not Found'
```

Could you try it out and lend me a hand? If not, I'll just go with bluebird. TBH, I'm not entirely sold on using `any-promise` vs. `bluebird`. Flexibility it's just not worth it in this case and inserts a non-deterministic behaviour which should be taken into account.

**Comment by nfantone** [comment]
Opened https://github.com/Foxandxss/koa-unless/pull/1 un `koa-unless`. Let's wait and see if we have any feedback.

**Comment by nfantone** [comment]
Added `"koa-unless": "nfantone/koa-unless#next"` as temporary dependency until my PR is merged.

**Comment by fl0w** [comment]
@nfantone I guess you're right. I'll take a look at what's up, we're using `jwt-then` so I'm not sure what's fails for you. Give me a few hours and I'll take a look at your code/test suite.

**Comment by fl0w** [comment]
I'm having a little trouble guessing why something fails, would it be possible to create a separate branch/commit with jwt-then, because my guess is you're not consuming the returned Promise correctly on check. The jwt is generated correctly w/ whatever you put in it - the audience check occurs in koa-jwt.

`var token = jwt.sign(...)` would return a promise, not a resolved value, so you'd have to append a `.then` and then do the suite test.

**Comment by nfantone** [comment]
@fl0w You mean lines likes this:

```js
  var token = jwt.sign({foo: 'bar'}, secret);
```

in `test.js`? I'm now using `jwt-then` there. Just plain `jsonwebtoken`.

**Comment by nfantone** [comment]
> would it be possible to create a separate branch/commit with jwt-then

Of course it would. You're very welcome to do it. I don't want this to come out wrong. Really. But I see no benefit in doing so.

>  because my guess is you're not consuming the returned Promise correctly on check. 

It might be, yes. But then, how would that account for bluebird working? Any ideas?

**Comment by stiang** [comment]
@flow @nfantone I feel that we shouldn’t let the Bluebird/jwt-then question sidetrack the issue at hand (supporting Koa v2). 

If @nfantone gets a working solution using Bluebird which solved the immediate problem, then that’s great! Others are of course welcome to improve on the branch later, for example by switching to jwt-then (if that is indeed an improvement - I don’t really know at this point).

**Comment by fl0w** [comment]
I'm on a phone and can't really do much :/ don't let it prolong, just super curious how u got one working but not the other (my guess is that by promisifying all, you also promisify sync calls, which is what you're actually calling - which is bad/wrong)

**Comment by nfantone** [comment]
@fl0w It most certainly is not related to `promisyAll`.

You can also try doing this:
```js
const jsonwebtoken = require('jsonwebtoken');
const verify       = Promise.promisify(require('jsonwebtoken').verify, { context: jsonwebtoken });
```

It'll work fine.

**Comment by nfantone** [comment]
@stiang All done over here! Just removed the "[WIP]" prefix.

**Comment by stiang** [comment]
@nfantone Awesome job! Will try to take a closer look tomorrow and push a release then.

**Comment by stiang** [comment]
@nfantone Sorry about the delay, it should be published as `koa-jwt@koa2` now. I also bumped the version to 2.0.0 for this branch, since the API has changed. 

Please let me know if there are any issues! I still haven’t gotten around to using koa2 myself, so this is not well tested by me. 

Btw, does anybody know why the throw "leaks through" to the output for the *should throw if secret neither provided by options or middleware* test, but not the rest? I couldn’t figure it out, and it would be nice if the test output remained clean.

**Comment by nfantone** [comment]
@stiang What exactly do you mean by "leaks through"?

If there's no `secret` defined, these lines:

```javascript
    secret = (ctx.state && ctx.state.secret) ? ctx.state.secret : opts.secret;
    if (!secret) {
      ctx.throw(500, 'Invalid secret\n');
    }
```

Will throw a 500. Maybe 500 is not the best code here? But, apart from that, LGTM.

**Comment by stiang** [comment]
Yeah, bad choice of words. I just mean that the test run displays the error and the stacktrace, like this:

test-that-leaks-error

But if I comment out that particular test it looks nice and clean, even though there should be lots of "throwing" going on behind the scenes:

test-that-is-normal

**Comment by stiang** [comment]
Nevermind, figured it out. It was due to using the 500 status code and not adding `{expose: true}` to the throw call. I switched to 401, which doesn’t require the option and is arguably more correct.

Thanks to @snooc for pointing it out in #36.

**Pull request merged**